src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/MethodResolver.cs

```diff
@@ -0,0 +1,127 @@
+
+            public override bool Equals(object? obj)
+            {
+                return Equals(obj as DeclaringTypeDescription);
+            }
+
+            public bool Equals(DeclaringTypeDescription? other)
+            {
+                return other is not null &&
+                       ModuleName == other.ModuleName &&
+                       ClassName == other.ClassName;
+            }
+
+            public override int GetHashCode()
+            {
+                return HashCode.Combine(ModuleName, ClassName);
+            }
+        }
+
+        private readonly Dictionary<string, List<Module>> _dllNameToModules = new(StringComparer.InvariantCultureIgnoreCase);
+        private readonly Dictionary<DeclaringTypeDescription, List<MethodInfo>> _declaringTypeToMethods = new();
+
+        public MethodResolver()
+        {
+            // Build a lookup table of all viable dll names to their backing reflection Module.
+            IEnumerable<Assembly> assemblies = AppDomain.CurrentDomain.GetAssemblies().Where(
+                assembly => !assembly.ReflectionOnly &&
+                !assembly.IsDynamic);
+
+            foreach (Assembly assembly in assemblies)
+            {
+                foreach (Module module in assembly.GetModules())
+                {
+                    if (!_dllNameToModules.TryGetValue(module.Name, out List<Module>? moduleList))
+                    {
+                        moduleList = new List<Module>();
+                        _dllNameToModules[module.Name] = moduleList;
+                    }
+
+                    moduleList.Add(module);
+                }
+            }
+        }
+
+        public List<MethodInfo> ResolveMethodDescription(MethodDescription methodDescription)
+        {
+            // A single description can match multiple MethodInfos
+            List<MethodInfo> matchingMethods = new();
+
+            List<MethodInfo> possibleMethods = GetMethodsForDeclaringType(methodDescription);
+            foreach (MethodInfo method in possibleMethods)
+            {
+                if (method.Name == methodDescription.MethodName)
+                {
+                    matchingMethods.Add(method);
+                }
+            }
+
+            return matchingMethods;
+        }
+
+        private List<MethodInfo> GetMethodsForDeclaringType(MethodDescription methodDescription)
+        {
+            // Maintain a cache for all methods for a given module+class.
+            DeclaringTypeDescription declType = new()
+            {
+                ModuleName = methodDescription.ModuleName,
+                ClassName = methodDescription.ClassName
+            };
+
+            if (_declaringTypeToMethods.TryGetValue(declType, out List<MethodInfo>? methods))
+            {
+                return methods;
+            }
+
+            List<MethodInfo> classMethods = new();
+            if (!_dllNameToModules.TryGetValue(methodDescription.ModuleName, out List<Module>? possibleModules))
```

Comment:
nit: Not necessarily a recommendation, just noticed some duplication since this `if` block is the same as what we have at the bottom of this method - the `foreach` loop will only execute if `possibleModules` contains elements, which suggests we could do away with this conditional and just let everything return at the bottom.